<a href="https://colab.research.google.com/github/BigBubl1k/data_science/blob/main/Yandex_market_%D0%A2%D1%8E%D1%80%D0%B8%D0%BD_%D0%98%D0%B3%D0%BE%D1%80%D1%8C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#импортируем библиотеки и загрузим данные в data_json

In [74]:
from urllib.request import urlopen
import json
url = 'http://5.181.109.139/marketer/api/data.json'
response = urlopen(url)
data_json = json.loads(response.read())


#Проверим типы данных и посмотрим на их запись

In [75]:
len(data_json)

2

In [76]:
type(data_json)

dict

In [77]:
type(data_json['items'])

list

In [78]:
data_json['items'][0]

{'competitor price': 1799,
 'competitor shop': 'VeloSmile',
 'original category': 'Спасательные жилеты и круги',
 'original id': '2094028',
 'original price': 3210,
 'original title': '"Cпасательный жилет ""Докер"" до 80 кг р.44-46"',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=aprice',
 'sima-land.ru price': 1955,
 'sima-land.ru updated': '2022-04-19 01:00:10'}

#Сделаем цикл, который проходится по каждому объекту словаря

Первый цикл пройдем вручную

In [79]:
article = data_json['items'][0]['original id']                      #артикул товара
whole_sale_price = data_json['items'][0]['sima-land.ru price']      #оптовая цена этого товара
original_price = data_json['items'][0]['original price']            #наша цена до преобразования
min_competitor_price = data_json['items'][0]['competitor price']    #минимальная цена конкурента на этот товар
our_price = (50+whole_sale_price)*100/55                            #расчет по формуле нашей цены товара

if our_price < min_competitor_price:                                #если наша цена окажется ниже миинимальной цены конкурентов, устанавливаем ее
  data_json['items'][item]['original price'] = min_competitor_price*0.97
data_json['items'][0]

{'competitor price': 1799,
 'competitor shop': 'VeloSmile',
 'original category': 'Спасательные жилеты и круги',
 'original id': '2094028',
 'original price': 3210,
 'original title': '"Cпасательный жилет ""Докер"" до 80 кг р.44-46"',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=aprice',
 'sima-land.ru price': 1955,
 'sima-land.ru updated': '2022-04-19 01:00:10'}

Теперь пройдемся по всем элементам

In [80]:
for item in data_json['items']:                       #проходимся по каждому объекту

  if item['competitor shop'] == 'Русэкспресс':        #если самая низкая цена конкурента от Русэкспресса,
    continue                                          #то пропускаем шаг цикла, т.к это наш магазин


  whole_sale_price = item['sima-land.ru price']       #оптовая цена товара
  min_competitor_price = item['competitor price']     #минимальная цена конкурента на этот товар


  if item['original id'] == article:                  #Так как нам нужен только первый набор данных по артиклу, 
    item['original price'] = original_price
    continue                                          #если артикул повторяет артикул прошлого цикла, пропускаем шаг

  else:
    article = item['original id']                     #если артикул новый, то переменной article присваем новое значение артикула
    our_price = (50+whole_sale_price)*100/55          #и считаем нашу цену по формуле

    original_price = item['original price']           #чтобы для других наборов данных с одинаковым артикулом также изменить
                                                      #цену на новую, если изменения будут, используем переменную original_price

  if our_price < min_competitor_price:                #если эта цена окажется ниже мин. цены конкурента, заменяем в данных ее
      item['original price'] = min_competitor_price*0.97
      original_price = min_competitor_price*0.97

  else:
    continue




In [81]:
#327100 записей всего

In [82]:
type(data_json['items'][0]['competitor price'])

int

In [83]:
data_json['items'][6]

{'competitor price': 1084,
 'competitor shop': 'Qwen',
 'original category': 'Игрушечные машинки и техника',
 'original id': '7517061',
 'original price': 1051.48,
 'original title': '100 секретных окошек "Транспорт',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/1699704981/offers?lr=213&sku=101625914255&how=aprice',
 'sima-land.ru price': 518,
 'sima-land.ru updated': '2022-04-19 01:00:11'}

In [84]:
data_json['items'][7]

{'competitor price': 1084,
 'competitor shop': 'Qwen',
 'original category': 'Игрушечные машинки и техника',
 'original id': '7517061',
 'original price': 1051.48,
 'original title': '100 секретных окошек "Транспорт',
 'original updated': '2022-04-18 00:37:18',
 'original url': 'https://market.yandex.ru/product/1699704981/offers?lr=213&sku=101625914255&how=aprice',
 'sima-land.ru price': 518,
 'sima-land.ru updated': '2022-04-19 01:00:11'}

Проверим, что для тех наборов данных, где мы изменили цену, цена изменилась для всех наборов с тем же артикулом

In [85]:
print('original price:',data_json['items'][6]['original price'],';','original price:',data_json['items'][7]['original price'])

original price: 1051.48 ; original price: 1051.48


In [86]:
for item in data_json['items'][6]:
  print(item ,type(item))

original id <class 'str'>
original title <class 'str'>
original price <class 'str'>
original category <class 'str'>
original url <class 'str'>
original updated <class 'str'>
sima-land.ru price <class 'str'>
sima-land.ru updated <class 'str'>
competitor shop <class 'str'>
competitor price <class 'str'>


Как видно по строке original price, цена скорретировалась (была изначально 1051), т.е, за счет того, что наша цена оказалась изначально ниже самой низкой цены конкурента, нам удалось поднять цену на товар на 48 коп.

#Теперь сформируем xml файл

In [91]:
!pip install dicttoxml

  Created wheel for dicttoxml: filename=dicttoxml-1.7.4-py3-none-any.whl size=17464 sha256=592fa37c5faf34d0b1e08b72b37b591aa8b78dbc3f051573965fefb835d4f824
  Stored in directory: /root/.cache/pip/wheels/18/44/5a/c7b93703f611ae3b50820efbaf265413e6b183a56cbec6f661
Successfully built dicttoxml


In [94]:
import dicttoxml
xml = dicttoxml.dicttoxml(data_json)

In [97]:
type(xml)

bytes

In [100]:
xml[:1000]

b'<?xml version="1.0" encoding="UTF-8" ?><root><items type="list"><item type="dict"><original_id type="str">2094028</original_id><original_title type="str">&quot;C\xd0\xbf\xd0\xb0\xd1\x81\xd0\xb0\xd1\x82\xd0\xb5\xd0\xbb\xd1\x8c\xd0\xbd\xd1\x8b\xd0\xb9 \xd0\xb6\xd0\xb8\xd0\xbb\xd0\xb5\xd1\x82 &quot;&quot;\xd0\x94\xd0\xbe\xd0\xba\xd0\xb5\xd1\x80&quot;&quot; \xd0\xb4\xd0\xbe 80 \xd0\xba\xd0\xb3 \xd1\x80.44-46&quot;</original_title><original_price type="int">3210</original_price><original_category type="str">\xd0\xa1\xd0\xbf\xd0\xb0\xd1\x81\xd0\xb0\xd1\x82\xd0\xb5\xd0\xbb\xd1\x8c\xd0\xbd\xd1\x8b\xd0\xb5 \xd0\xb6\xd0\xb8\xd0\xbb\xd0\xb5\xd1\x82\xd1\x8b \xd0\xb8 \xd0\xba\xd1\x80\xd1\x83\xd0\xb3\xd0\xb8</original_category><original_url type="str">https://market.yandex.ru/product/843306581/offers?lr=213&amp;sku=101192028615&amp;how=aprice</original_url><original_updated type="str">2022-04-18 00:37:18</original_updated><sima-land.ru_price type="int">1955</sima-land.ru_price><sima-land.ru_update

In [109]:
import xml.etree.ElementTree as ET
tree = ET.ElementTree(ET.fromstring(xml))

In [123]:
tree.write('yafbs_prices_updated.xml')

In [124]:
type(tree)

xml.etree.ElementTree.ElementTree

#На этом задача выполнена

#Посмотрим на альтернативные способы решения)

сформируем функцией вид xml файла, но в формате string

In [104]:
def json2xml(json_obj, line_padding=""):
    result_list = list()

    json_obj_type = type(json_obj)

    if json_obj_type is list:
        for sub_elem in json_obj:
            result_list.append(json2xml(sub_elem, line_padding))

        return "\n".join(result_list)

    if json_obj_type is dict:
        for tag_name in json_obj:
            sub_obj = json_obj[tag_name]
            result_list.append("%s<%s>" % (line_padding, tag_name))
            result_list.append(json2xml(sub_obj, "\t" + line_padding))
            result_list.append("%s</%s>" % (line_padding, tag_name))

        return "\n".join(result_list)

    return "%s%s" % (line_padding, json_obj)

In [105]:
result = json2xml(data_json)
print(result[:5000])

<items>
	<original id>
		2094028
	</original id>
	<original title>
		"Cпасательный жилет ""Докер"" до 80 кг р.44-46"
	</original title>
	<original price>
		3210
	</original price>
	<original category>
		Спасательные жилеты и круги
	</original category>
	<original url>
		https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=aprice
	</original url>
	<original updated>
		2022-04-18 00:37:18
	</original updated>
	<sima-land.ru price>
		1955
	</sima-land.ru price>
	<sima-land.ru updated>
		2022-04-19 01:00:10
	</sima-land.ru updated>
	<competitor shop>
		VeloSmile
	</competitor shop>
	<competitor price>
		1799
	</competitor price>
	<original id>
		2094028
	</original id>
	<original title>
		"Cпасательный жилет ""Докер"" до 80 кг р.44-46"
	</original title>
	<original price>
		3210
	</original price>
	<original category>
		Спасательные жилеты и круги
	</original category>
	<original url>
		https://market.yandex.ru/product/843306581/offers?lr=213&sku=101192028615&how=a

In [106]:
print(result[-150:])

ed>
	<competitor shop>
		kv.mart
	</competitor shop>
	<competitor price>
		319
	</competitor price>
</items>
<updated>
	2022-04-20 06:00:08
</updated>


In [36]:
for i in data_json:
  print(i)

items
updated


In [45]:
data_json['updated']

'2022-04-20 06:00:08'

Выглядит довольно хорошо, нужно только сохранить в xml (не нашел способ))

Попробуем вручную прописать деревья

In [63]:
import xml.etree.cElementTree as ET

In [66]:
root = ET.Element('root')

#ET.SubElement(root, 'items').text = data_json['items']

items = ET.SubElement(root, "items")

for i in data_json['items']:

  ET.SubElement(items, 'original id').text = i['original id']
  ET.SubElement(items, 'original title').text = i['original title']
  ET.SubElement(items, 'original price').text = str(i['original price'])
  ET.SubElement(items, 'original category').text = i['original category']
  ET.SubElement(items, 'original url').text = i['original url']
  ET.SubElement(items, 'original updated').text = i['original updated']
  ET.SubElement(items, 'sima-land.ru price').text = str(i['sima-land.ru price'])
  ET.SubElement(items, 'sima-land.ru updated').text = i['sima-land.ru updated']
  ET.SubElement(items, 'competitor shop').text = i['competitor shop']
  ET.SubElement(items, 'competitor price').text = str(i['competitor price'])

updated = ET.SubElement(root, "updated").text = data_json['updated']

tree = ET.ElementTree(root)
tree.write('yafbs_prices_updated1.xml')


In [67]:
type(tree)


xml.etree.ElementTree.ElementTree

In [68]:
import xml.etree.ElementTree as ET
from xml.etree.ElementTree import tostring
tree=ET.parse('yafbs_prices_updated1.xml')
tree = tree.getroot()
t = tostring(tree)
t = t.lower()
tree= ET.fromstring(t)

ParseError: ignored

In [61]:
with open("yafbs_prices_updated.xml", "w") as f:
    f.write(ET.tostring(tree))

AttributeError: ignored

Сталкиваюсь с непонятными ошибками)